### PR TITLE
added check for parameter description

### PR DESCRIPTION
**Pull request by KevinMenden**
Adding a lint check to look missing parameter descriptions in the schema and warn about parameters outside of groups.

## PR checklist

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` is updated
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated

### CODE REVIEW

**Comment by KevinMenden** [comment]
closes #1017

File: nf_core/lint/schema_description.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    self.schema_obj.get_schema_path(self.wf_path)
+    self.schema_obj.get_wf_params()
+    self.schema_obj.no_prompts = True
+    self.schema_obj.load_lint_schema()
+
+    # Get ungrouped params
+    if "properties" in self.schema_obj.schema.keys():
+        ungrouped_params = self.schema_obj.schema["properties"].keys()
+        for up in ungrouped_params:
+            warned.append(f"Ungrouped param in schema {up}")
```
Comment (ewels) [inline nf_core/lint/schema_description.py:32]:
```suggestion
            warned.append(f"Ungrouped param in schema: `{up}`")
```

File: nf_core/lint/schema_description.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Iterate over groups and add warning for parameters without a description
+    for group_key in self.schema_obj.schema["definitions"].keys():
+        group = self.schema_obj.schema["definitions"][group_key]
+        for param_key in group["properties"].keys():
+            param = group["properties"][param_key]
+            if not "description" in param.keys():
+                warned.append(f"No description provided in schema for parameter '{param_key}'")
```
Comment (ewels) [inline nf_core/lint/schema_description.py:40]:
```suggestion
                warned.append(f"No description provided in schema for parameter: `{param_key}`")
```

File: nf_core/lint/schema_description.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    if "properties" in self.schema_obj.schema.keys():
+        ungrouped_params = self.schema_obj.schema["properties"].keys()
+        for up in ungrouped_params:
+            warned.append(f"Ungrouped param in schema {up}")
+
+    # Iterate over groups and add warning for parameters without a description
+    for group_key in self.schema_obj.schema["definitions"].keys():
+        group = self.schema_obj.schema["definitions"][group_key]
+        for param_key in group["properties"].keys():
+            param = group["properties"][param_key]
```
Comment (ewels) [inline nf_core/lint/schema_description.py:38]:
```suggestion
        for param_key, param in group["properties"].items():
```

File: nf_core/lint/schema_description.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            if not "description" in param.keys():
```
Comment (ewels) [inline nf_core/lint/schema_description.py:39]: Both valid, but I find this easier to read:

```suggestion
            if "description" not in param.keys():
```

**Review by ewels — CHANGES_REQUESTED** [review]
Looks good, but please add some code so that params can be ignored using the lint config file.

**Comment by KevinMenden** [comment]
Ah sure, sorry! Will add that.
Thanks for the review! :+1:

File: nf_core/lint/schema_description.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    # Get parameters that should be ignored according to the linting config
+    ignore_params = self.lint_config.get("schema_description", [])
+
+    # Get ungrouped params
+    if "properties" in self.schema_obj.schema.keys():
+        ungrouped_params = self.schema_obj.schema["properties"].keys()
+        for up in ungrouped_params:
+            if up in ignore_params:
+                continue
+            warned.append(f"Ungrouped param in schema: `{up}`")
```
Comment (ewels) [inline nf_core/lint/schema_description.py:38]: Almost - we need to report the tests that were ignored though..

```suggestion
            if up in ignore_params:
                ignored.append(f"Ignored ungrouped param in schema: `{up}`")
            else:
                warned.append(f"Ungrouped param in schema: `{up}`")
```

File: nf_core/lint/schema_description.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    # Iterate over groups and add warning for parameters without a description
+    for group_key in self.schema_obj.schema["definitions"].keys():
+        group = self.schema_obj.schema["definitions"][group_key]
+        for param_key, param in group["properties"].items():
+            if param_key in ignore_params:
+                continue
```
Comment (ewels) [inline nf_core/lint/schema_description.py:45]:
```suggestion
                ignored.append(f"Ignoring description check for param in schema: `{param_key}`")
                continue
```

File: nf_core/lint/schema_description.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    The ``nextflow_schema.json`` pipeline schema should describe every flat parameter
+    Furthermore warns about parameters outside of groups
+
+    * Warning: Parameters in ``nextflow_schema.json`` without a description
+    * Warning: Parameters in ``nextflow_schema.json`` that are defined outside of a group
+    """
+    passed = []
+    warned = []
+    failed = []
```
Comment (ewels) [inline nf_core/lint/schema_description.py:18]:
```suggestion
```

File: nf_core/lint/schema_description.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        for param_key, param in group["properties"].items():
+            if param_key in ignore_params:
+                continue
+            if "description" not in param.keys():
+                warned.append(f"No description provided in schema for parameter: `{param_key}`")
+
+    for ip in ignore_params:
+        ignored.append(f"Parameter is ignored: `{ip}`")
+
+    return {"passed": passed, "warned": warned, "failed": failed}
```
Comment (ewels) [inline nf_core/lint/schema_description.py:52]:
```suggestion
    return {"passed": passed, "warned": warned, "ignored": ignored}
```

**Review by ewels — CHANGES_REQUESTED** [review]
Need to report the ignored tests, then looks good 👍🏻

**Comment by KevinMenden** [comment]
:facepalm:

**Review by ewels — APPROVED** [review]
LGTM 🚀

**Pull request merged**